lessons/dotnetcore31/durable-functions/chaining/README.md

```diff
@@ -0,0 +1,463 @@
+
+   ```csharp
+   private const int FirstRetryIntervalInMilliseconds = 1000;
+   ```
+
+   3.2 Set the maximum number of attempts to 3.
+
+   ```csharp
+   private const int MaxNumberOfAttempts = 3;
+   ```
+
+   3.3 Set the maximum amount of time to wait in between retry attempts to 1000 ms.
+
+   ```csharp
+   private const int MaxRetryIntervalInMilliseconds = 1000;
+   ```
+
+   3.4 Set the maximum amount of time to spend doing retries to 7000 ms.
+
+   ```csharp
+   private const int RetryTimeoutInMilliseconds = 7000;
+   ```
+
+   > 📝 **Tip** - It is a best practice to inject the value via environment variables. When doing so make sure that the variable type is converted into a number.
+   > ❔ **Question** - How long would the Orchestrator Function retry failed calls by default?
+
+4. Create a instance of the retry options class and set the configuration values.
+
+   ```csharp
+   var retryOptions = new RetryOptions(
+      firstRetryInterval: TimeSpan.FromMilliseconds(FirstRetryIntervalInMilliseconds),
+      maxNumberOfAttempts: MaxNumberOfAttempts)
+   {
+      MaxRetryInterval = TimeSpan.FromMilliseconds(MaxRetryIntervalInMilliseconds),
+      RetryTimeout = TimeSpan.FromMilliseconds(RetryTimeoutInMilliseconds)
+   };
+   ```
+
+   > 🔎 **Observation** - The constructor of the retry options calls only expects a minimum configuration.
+
+5. Update the calls of the Activity Functions to use retries and consider the retry configuration.
+
+   ```csharp
+   var userName = await context.CallActivityWithRetryAsync<string>(
+      "GitHubInfoRetryOrchestrator_GetRepositoryDetailsByName",
+      retryOptions,
+      input);
+   ...
+   var userDetails = await context.CallActivityWithRetryAsync<string>(
+      "GitHubInfoRetryOrchestrator_GetUserDetailsByName",
+      retryOptions,
+      userName);
+   ```
+
+6. In order to enforce an error we will cheat a bit. When we call the function we will hand over a parameter called `raiseException` that will be evaluated in the activity function `GitHubInfoRetryOrchestrator_GetRepositoryDetailsByName`. In case that the parameter is set to `true` we will throw an error. Change the logic of the Activity Function and the caller accordingly:
+
+   ```csharp
+   ...
+   [FunctionName("GitHubInfoRetryOrchestrator_GetRepositoryDetailsByName")]
+   public static async Task<string> GetRepositoryDetails([ActivityTrigger] string[] inputs, ILogger log)
+   {
+      var name = inputs[0];
+      var raiseException = bool.Parse(inputs[1]);
+
+      if (raiseException)
+            throw new Exception("Stop!");
+
+      var githubClient = new GitHubClient(new ProductHeaderValue("azure-functions-university"));
+
+      var request = new SearchRepositoriesRequest(name);
+
+      var searchResult = await githubClient.Search.SearchRepo(request);
+      var repository = searchResult.Items.Single(x => x.Name == name);
+
+      return repository.Owner.Login;
+   }
+   ...
+   ```
+
+   ```csharp
```

Comment:
I suggest to be very explicit here and mention the name of the ochestrator function, and the place where this code needs to be put.